scanner/src/funTest/kotlin/experimental/AbstractProvenanceBasedStorageFunTest.kt

```diff
@@ -0,0 +1,239 @@
+            }
+
+            "fail if provenance information is missing" {
+                val scanResult = createScanResult(provenance = UnknownProvenance)
+
+                shouldThrow<ScanStorageException> { storage.write(scanResult) }
+            }
+
+            "fail if a result for the same provenance and scanner is already stored" {
+                val scanResult = createScanResult()
+
+                storage.write(scanResult)
+                shouldThrow<ScanStorageException> { storage.write(scanResult) }
+            }
+
+            "fail if the provenance contains a VCS path" {
+                val scanResult = createScanResult(
+                    provenance = createRepositoryProvenance(
+                        vcsInfo = VcsInfo.valid().copy(path = "path")
+                    )
+                )
+
+                shouldThrow<ScanStorageException> { storage.write(scanResult) }
+            }
+
+            "ignore the VCS revision in favor of the resolved revision when adding a duplicate" {
+                val scanResult1 = createScanResult(provenance = createRepositoryProvenance())
+                val scanResult2 = createScanResult(
+                    provenance = createRepositoryProvenance(
+                        vcsInfo = VcsInfo.valid().copy(revision = "another")
+                    )
+                )
+
+                storage.write(scanResult1)
+                shouldThrow<ScanStorageException> { storage.write(scanResult2) }
+            }
+        }
+
+        "Reading a scan result" should {
+            "find all scan results for a provenance" {
+                val scanResult1 = createScanResult()
+                val scanResult2 = createScanResult(scannerDetails = createScannerDetails(name = "other"))
+
+                storage.write(scanResult1)
+                storage.write(scanResult2)
+
+                val readResult = storage.read(scanResult1.provenance as KnownProvenance)
+
+                readResult should containExactlyInAnyOrder(scanResult1, scanResult2)
+            }
+
+            "use only the resolved revision for matching and return the provided provenance" {
+                val provenance1 = createRepositoryProvenance()
+                val provenance2 = provenance1.copy(vcsInfo = provenance1.vcsInfo.copy(revision = "another"))
+
+                val scanResult = createScanResult(provenance = provenance1)
+
+                storage.write(scanResult)
+
+                val readResult = storage.read(provenance2)
+
+                readResult should containExactlyInAnyOrder(scanResult.copy(provenance = provenance2))
+            }
+
+            "fail if the provenance contains a VCS path" {
+                val provenance = createRepositoryProvenance(vcsInfo = VcsInfo.valid().copy(path = "path"))
+                val criteria = createScannerDetails().toCriteria()
+
+                shouldThrow<ScanStorageException> { storage.read(provenance) }
```

Comment:
Nit: The documentation for the `read` functions added in the previous commit does not mention under which circumstances a `ScanStorageException` is thrown.